File: src/specification.cpp

// specification.cpp

// Copyright 2004 Jeremy Bertram Maitin-Shepard.

// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License Version
// 2 as published by the Free Software Foundation.

// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
// USA

#include "specification.hpp"

// Used by Specification::createConferenceLists()
#include <map>
#include <algorithm>

namespace cws
{

  Specification::Specification(int participantCount,
                               int firstGroupCount,
                               int slotCount)
    : participantCount_(participantCount),
      slotCount_(slotCount),
      firstGroupCount_(firstGroupCount),
      secondGroupCount_(participantCount - firstGroupCount),
      compatibleSlots_(participantCount_ * participantCount_),
      validSlot_(participantCount * slotCount),
      validSlots_(participantCount),
      conferenceList_(participantCount),
      conferenceValue_(firstGroupCount_ * secondGroupCount_),
      conferenceCostList_(firstGroupCount_),
      travelCost_(firstGroupCount * firstGroupCount),
      maxConsecutive_(participantCount),
      slotTime_(slotCount),
      slotDuration_(0),
      conferenceValueWeight_(0),
      waitWeight_(0),
      unsatisfiedPreferenceWeight_(0),
      travelWeight_(0),
      consecutiveWeight_(0)
  {}

  Specification::Specification()
    : participantCount_(0),
      slotCount_(0),
      firstGroupCount_(0),
      secondGroupCount_(0),
      slotDuration_(0),
      conferenceValueWeight_(0),
      waitWeight_(0),
      unsatisfiedPreferenceWeight_(0),
      travelWeight_(0),
      consecutiveWeight_(0)
  {}

  void Specification::setSize(int participantCount,
                              int firstGroupCount,
                              int slotCount)
  {
    participantCount_ = participantCount;
    slotCount_ = slotCount;
    firstGroupCount_ = firstGroupCount;
    secondGroupCount_ = participantCount - firstGroupCount;
    compatibleSlots_.initialize(participantCount_ * participantCount_);
    validSlot_.initialize(participantCount_ * slotCount_);
    validSlots_.initialize(participantCount_);
    conferenceList_.initialize(participantCount_);
    conferenceValue_.initialize(firstGroupCount_ * secondGroupCount_);
    conferenceCostList_.initialize(firstGroupCount_);
    travelCost_.initialize(firstGroupCount * firstGroupCount);
    slotTime_.initialize(slotCount);
    maxConsecutive_.initialize(participantCount);
  }

  void Specification::createCache()
  {
    createConferenceLists();
    createConferenceCostList();
    createSlotDataCache();
  }

  void Specification::createSlotDataCache()
  {
    // First generate valid slot cache
    for (int p = 0; p < participantCount(); ++p)
    {
      // Set validSlot to false for all slots
      for (int slot = 0; slot < slotCount(); ++slot)
        validSlot_[p + slot * participantCount()] = false;

      for (int i = 0; i < validSlots(p).size(); ++i)
        validSlot_[p + validSlots(p)[i] * participantCount()] = true;
    }

    // Then generate compatible slot lists
    for (int p1 = 0; p1 < participantCount(); ++p1)
    {
      for (int p2 = 0; p2 < participantCount(); ++p2)
      {
        int count = 0;
        for (int i = 0; i < validSlots(p1).size(); ++i)
        {
          if (validSlot(p2, validSlots(p1)[i]))
            ++count;
        }

        compatibleSlots_[p1 + p2 * participantCount()]
          .initialize(count);
        count = 0;

        for (int i = 0; i < validSlots(p1).size(); ++i)
        {
          int slot = validSlots(p1)[i];
          if (validSlot(p2, slot))
          {
            compatibleSlots_[p1 + p2 * participantCount()][count]
              = slot;
            ++count;
          }
        }
      }
    }
  }

  void Specification::createConferenceLists()
  {
    for (int p1 = 0; p1 < firstGroupCount(); ++p1)
    {
      typedef std::multimap<int, std::pair<int, int> > MapType;
      std::multimap<int, std::pair<int, int> > map;
      for (int p2 = 0; p2 < secondGroupCount(); ++p2)
      {
        for (int i = 0; i < maximumConferences(p1, p2); ++i)
        {
          int value = conferenceValue(p1, p2, i);
          map.insert
            (std::make_pair
             (value, std::make_pair(p2 + firstGroupCount(), i)));
        }
      }

      conferenceList_[p1].initialize((int)map.size());
      int count = 0;
      for (MapType::const_iterator it = map.begin(),
             end = map.end();
           it != end;
           ++it)
      {
        conferenceList_[p1][count] = it->second;
        ++count;
      }
    }

    for (int p2 = 0; p2 < secondGroupCount(); ++p2)
    {
      typedef std::multimap<int, std::pair<int, int> > MapType;
      std::multimap<int, std::pair<int, int> > map;
      for (int p1 = 0; p1 < firstGroupCount(); ++p1)
      {
        for (int i = 0; i < maximumConferences(p1, p2); ++i)
        {
          int value = conferenceValue(p1, p2, i);
          map.insert(std::make_pair(value, std::make_pair(p1, i)));
        }
      }

      conferenceList_[p2 + firstGroupCount()]
        .initialize((int)map.size());
      
      int count = 0;
      for (MapType::const_iterator it = map.begin(),
             end = map.end();
           it != end;
           ++it)
      {
        conferenceList_[p2 + firstGroupCount()][count] = it->second;
        ++count;
      }
    }
  }

  void Specification::createConferenceCostList()
  {
    for (int p1 = 0; p1 < firstGroupCount(); ++p1)
    {
      int count = 0;

      // First determine the number of participants in the second
      // group for which a conference with `p1' is valuable.
      for (int p2 = 0; p2 < secondGroupCount(); ++p2)
      {
        if (maximumConferences(p1, p2) > 0)
          ++count;
      }

      conferenceCostList_[p1].initialize(count);
      count = 0;

      for (int p2 = 0; p2 < secondGroupCount(); ++p2)
      {
        int maxConferences = maximumConferences(p1, p2);
        if (maxConferences == 0)
          continue;

        conferenceCostList_[p1][count].first = p2;
        conferenceCostList_[p1][count]
          .second.initialize(maxConferences + 1);

        conferenceCostList_[p1][count].second[maxConferences] = 0;

        int value = 0;
        for (int i = maxConferences - 1; i >= 0; --i)
        {
          value += conferenceValue(p1, p2, i);
          conferenceCostList_[p1][count].second[i] = value;
        }
        
        ++count;
      }
    }
  }
  
} // namespace cws

File: lib/org/commschool/scheduler/Specification.java

// Specification.java

// Copyright 2004 Jeremy Bertram Maitin-Shepard.

// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License Version
// 2 as published by the Free Software Foundation.

// This program is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
// USA

package org.commschool.scheduler;

import java.io.*;

public final class Specification
{
  private int slotDuration;
  private float conferenceValueWeight;
  private float waitWeight;
  private float unsatisfiedPreferenceWeight;
  private float travelWeight;
  private float consecutiveWeight;
    
  private int participantCount;
  private int slotCount;
  private int firstGroupCount, secondGroupCount;

  private int[][] validSlots;
  private int[][] conferenceValue;
  private int[] travelCost;
  private int[] slotTime;
  private int[] maxConsecutive;

  public Specification(int participantCount,
                       int firstGroupCount,
                       int slotCount)
  {
    setSize(participantCount, firstGroupCount, slotCount);
  }

  public Specification() {}

  public int conferenceValue(int p1, int p2, int existingConferences)
  {
    return conferenceValue[p1 + p2 * firstGroupCount()]
      [existingConferences];
  }

  public int maximumConferences(int p1, int p2)
  {
    return conferenceValue[p1 + p2 * firstGroupCount()].length;
  }

  public int maximumConsecutive(int p)
  {
    return maxConsecutive[p];
  }

  public int[] validSlots(int p1)
  {
    return validSlots[p1];
  }

  public int travelCost(int p1, int p2)
  {
    return travelCost[p1 + p2 * firstGroupCount()];
  }

  public int slotTime(int slot)
  {
    return slotTime[slot];
  }

  public int[][] validSlotData() {return validSlots;}
  public int[][] conferenceValueData() {return conferenceValue;}
  public int[] travelCostData() {return travelCost;}
  public int[] maximumConsecutiveData() {return maxConsecutive;}
  public int[] slotTimeData() {return slotTime;}

  public void setSize(int participantCount,
                      int firstGroupCount,
                      int slotCount)
  {
    this.participantCount = participantCount;
    this.firstGroupCount = firstGroupCount;
    this.secondGroupCount = participantCount - firstGroupCount;
    this.slotCount = slotCount;

    validSlots = new int[participantCount][];
    conferenceValue = new int[firstGroupCount * secondGroupCount][];
    travelCost = new int[firstGroupCount * firstGroupCount];
    slotTime = new int[slotCount];
    maxConsecutive = new int[participantCount];
  }

  public int participantCount() {return participantCount;}
  public int firstGroupCount() {return firstGroupCount;}
  public int secondGroupCount() {return secondGroupCount;}
  public int slotCount() {return slotCount;}
  public int slotDuration() {return slotDuration;}
  public float conferenceValueWeight() {return conferenceValueWeight;}
  public float waitWeight() {return waitWeight;}
  public float unsatisfiedPreferenceWeight() {return unsatisfiedPreferenceWeight;}
  public float travelWeight() {return travelWeight;}
  public float consecutiveWeight() {return consecutiveWeight;}

  public void setSlotDuration(int slotDuration)
  {
    this.slotDuration = slotDuration;
  }

  public void setConferenceValueWeight(float conferenceValueWeight)
  {
    this.conferenceValueWeight = conferenceValueWeight;
  }

  public void setWaitWeight(float waitWeight)
  {
    this.waitWeight = waitWeight;
  }

  public void setUnsatisfiedPreferenceWeight(float unsatisfiedPreferenceWeight)
  {
    this.unsatisfiedPreferenceWeight = unsatisfiedPreferenceWeight;
  }

  public void setTravelWeight(float travelWeight)
  {
    this.travelWeight = travelWeight;
  }

  public void setConsecutiveWeight(float consecutiveWeight)
  {
    this.consecutiveWeight = consecutiveWeight;
  }

  public void write(Writer writer) throws IOException
  {
    // Write general parameters
    writer.write(String.valueOf(participantCount));
    writer.write(' ');
    writer.write(String.valueOf(firstGroupCount));
    writer.write(' ');
    writer.write(String.valueOf(slotCount));
    writer.write(' ');
    writer.write(String.valueOf(slotDuration));
    writer.write(' ');
    writer.write(String.valueOf(conferenceValueWeight));
    writer.write(' ');
    writer.write(String.valueOf(waitWeight));
    writer.write(' ');
    writer.write(String.valueOf(unsatisfiedPreferenceWeight));
    writer.write(' ');
    writer.write(String.valueOf(travelWeight));
    writer.write(' ');
    writer.write(String.valueOf(consecutiveWeight));
    writer.write('\n');

    // Write slotTimeData
    for (int i = 0; i < slotCount; ++i)
    {
      if (i != 0)
        writer.write(' ');
      writer.write(String.valueOf(slotTime(i)));
    }
    writer.write('\n');

    // Write travelCostData
    for (int p1 = 0; p1 < firstGroupCount(); ++p1)
    {
      for (int p2 = 0; p2 < firstGroupCount(); ++p2)
      {
        if (p1 != 0 || p2 != 0)
          writer.write(' ');
        writer.write(String.valueOf(travelCost(p2, p1)));
      }
    }
    writer.write('\n');

    // Write maximumConsecutiveData
    for (int p = 0; p < participantCount(); ++p)
    {
      if (p != 0)
        writer.write(' ');
      writer.write(String.valueOf(maximumConsecutive(p)));
    }
    writer.write('\n');

    // Write validSlotData
    for (int p = 0; p < participantCount(); ++p)
    {
      writer.write(String.valueOf(validSlots(p).length));
      for (int i = 0; i < validSlots(p).length; ++i)
      {
        writer.write(' ');
        writer.write(String.valueOf(validSlots(p)[i]));
      }
      writer.write('\n');
    }

    // Write conferenceValueData
    for (int p2 = 0; p2 < secondGroupCount(); ++p2)
    {
      for (int p1 = 0; p1 < firstGroupCount(); ++p1)
      {
        if (p1 != 0)
          writer.write("  ");
        writer.write(String.valueOf(maximumConferences(p1, p2)));
        for (int i = 0; i < maximumConferences(p1, p2); ++i)
        {
          writer.write(' ');
          writer.write(String.valueOf(conferenceValue(p1, p2, i)));
        }
      }
      writer.write('\n');
    }
  }

  public void read(TextDataReader reader) throws IOException
  {
    int participantCount, firstGroupCount, slotCount;
    int slotDuration;
    float conferenceValueWeight, waitWeight,
      unsatisfiedPreferenceWeight, travelWeight, consecutiveWeight;

    participantCount = reader.readInt();
    firstGroupCount = reader.readInt();
    slotCount = reader.readInt();
    slotDuration = reader.readInt();
    conferenceValueWeight = reader.readFloat();
    waitWeight = reader.readFloat();
    unsatisfiedPreferenceWeight = reader.readFloat();
    travelWeight = reader.readFloat();
    consecutiveWeight = reader.readFloat();

    if (participantCount < 0
        || firstGroupCount < 0
        || firstGroupCount < participantCount
        || slotDuration < 0
        || conferenceValueWeight < 0
        || waitWeight < 0
        || unsatisfiedPreferenceWeight < 0
        || travelWeight < 0
        || consecutiveWeight < 0)
    {
      throw new IOException("Integrity check failed");
    }

    setSize(participantCount, firstGroupCount, slotCount);
    setSlotDuration(slotDuration);
    setConferenceValueWeight(conferenceValueWeight);
    setWaitWeight(waitWeight);
    setUnsatisfiedPreferenceWeight(unsatisfiedPreferenceWeight);
    setTravelWeight(travelWeight);
    setConsecutiveWeight(consecutiveWeight);
    
    // Read slotTimeData
    for (int i = 0; i < slotCount; ++i)
    {
      slotTime[i] = reader.readInt();
    }

    // Read travelCostData
    for (int i = 0;
         i < firstGroupCount() * secondGroupCount();
         ++i)
    {
      int value = reader.readInt();
      if (value < 0)
      {
        throw new IOException("Invalid travel cost: " + value);
      }
      travelCost[i] = value;
    }

    // Read maximumConsecutiveData
    for (int p = 0; p < participantCount(); ++p)
    {
      int value = reader.readInt();
      if (value < 0)
        throw new IOException("Invalid maximum number of consecutive conferences: " + value);
      maxConsecutive[p] = value;
    }

    // Read validSlotData
    for (int p = 0; p < participantCount(); ++p)
    {
      int length = reader.readInt();;
      if (length < 0)
      {
        throw new IOException("Invalid length: " + length);
      }
      
      validSlots[p] = new int[length];

      for (int i = 0; i < length; ++i)
      {
        int value = reader.readInt();
        if (value < 0 || value >= slotCount)
          throw new IOException("Invalid slot: " + value);
        validSlots[p][i] = value;
      }
    }

    // Read conferenceValueData
    for (int i = 0;
         i < firstGroupCount() * secondGroupCount();
         ++i)
    {
      int length = reader.readInt();;
      if (length < 0)
      {
        throw new IOException("Invalid length: " + length);
      }
      
      conferenceValue[i] = new int[length];

      for (int j = 0; i < length; ++j)
      {
        int value = reader.readInt();
        if (value < 0)
        {
          throw new IOException("Invalid value: " + value);
        }
        conferenceValue[i][j] = value;
      }
    }
  }

  
}
